zephyr/lib/regions_mm.c

```diff
@@ -1,53 +1,339 @@
+			// czy moze zakladamy ze span bedzie tylko dla jednorodnych heapow
+			allocation_error_code = sys_mem_blocks_alloc_contiguous(
+				heap->physical_blocks_allocators[i],
+				blck_size,
+				(void*)ptr);
+		}
+		else if(blck_size > alloc_size)
+		{
+			allocation_error_code = sys_mem_blocks_alloc(
+				heap->physical_blocks_allocators[i],
+				blck_size,
+				(void**)&ptr);
+		}
+		/* Save allocation data for purpose of freeing correctly afterwards
+		 * should be changed to a implementation in memblocks that keeps
+		 * alloc data there.
+		 */
+		if (allocation_error_code == 0)
+		{
+			new_allocation->allocation_ptr = (uint32_t*)ptr;
+			new_allocation->allocation_size = blck_size;
+		}
+	}
+
+	/* If ptr is NULL it means we did not allocate anything and we should break execution here */
+	if (ptr == NULL) return ptr;
+
+	/* Now we need to iterate over mapped physical memory check if we already have
+	 * this specific
+	 */
+	sys_snode_t *iterator;
+	bool space_already_mapped = 0;
+	SYS_SLIST_FOR_EACH_NODE(&physical_mappings, iterator)
+	{
+		struct active_physical_mapping *mapping = CONTAINER_OF(iterator,struct active_physical_mapping, node);
+
+		if(is_range_in_range(ptr, alloc_size, (uint8_t*)mapping->mapping_ptr , mapping->mapping_size))
+		{
+			space_already_mapped = 1;
+		}
+	}
+
+	/* If the space was not already mapped we map physical pages */
+	if (!space_already_mapped)
+	{
+		uint8_t *aligned_ptr = get_ptr_aligned_to_physical(ptr);
+		size_t aligned_alloc_size = get_size_for_ptr_aligned_to_physical(ptr, alloc_size);
+		sys_mm_drv_map_region(aligned_ptr, (uintptr_t)NULL, aligned_alloc_size, SYS_MM_MEM_PERM_RW);
+		struct active_physical_mapping *new_mapping = rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM, sizeof(struct active_physical_mapping));
+		new_mapping->mapping_ptr = (uint32_t*)aligned_ptr;
+		new_mapping->mapping_size = aligned_alloc_size;
+		sys_slist_append(&physical_mappings,&new_mapping->node);
+	}
+	
+	sys_slist_append(&heap->active_allocations, &new_allocation->node);
+
+	return ptr;
+}
+
+int free_heap(struct virtual_memory_heap* heap)
+{
+	if (! sys_slist_is_empty(&heap->active_allocations))
+	{
+		return -ENOTEMPTY;
+	}
+
+	for (size_t i = 0; i < MAXIMUM_BLOCK_SIZES; i++)
+	{
+		if (heap->physical_blocks_allocators[i] != 0)
+		{
+			rfree(heap->physical_blocks_allocators[i]->bitmap->bundles);
+			rfree(heap->physical_blocks_allocators[i]->bitmap);
+			rfree(heap->physical_blocks_allocators[i]);
+		}
+	}
+	rfree(heap);
+}
+
+int free_from_heap(uint8_t *ptr, struct virtual_memory_heap *heap)
+{
+	sys_snode_t *iterator;
+	struct active_allocation* allocation = NULL;
+	int retval = 0;
+	SYS_SLIST_FOR_EACH_NODE(&heap->active_allocations, iterator)
+	{
+		if (ptr == (uint8_t *)CONTAINER_OF(iterator, struct active_allocation, node)->allocation_ptr)
+		{
+			allocation = (struct active_allocation*)
+				CONTAINER_OF(iterator,struct active_allocation, node)->allocation_ptr;
```

Comment:
Can we move `allocation = ...` up before if? We won't need to call the CONTAINER_OF twice